simulator/infra/cache/cache_tag_array.h

```diff
@@ -1,82 +1,137 @@
+    public:
+        explicit LRUCacheInfo( std::size_t capacity)
+            : CAPACITY( capacity)
+        {
+            assert( capacity != 0u);
+
+            lru_hash.reserve( CAPACITY);
+        }
+
+        auto get_capacity() const { return CAPACITY; }
+        auto size() const { return number_of_elements; }
+        bool empty() const { return size() == 0u; }
+
+        void touch( std::size_t key);
+        std::size_t update();
+
+
+    private:
+        std::list<std::size_t> lru_list{};
+        std::unordered_map<std::size_t, typename std::list<std::size_t>::const_iterator> lru_hash{};
+
+        std::size_t number_of_elements = 0u;
+        std::size_t CAPACITY;
 };
 
-class CacheTagArrayCheck : private Log
+
+
+class LRUModule
 {
-protected:
-    CacheTagArrayCheck( uint32 size_in_bytes,
-                        uint32 ways,
-                        uint32 line_size,
-                        uint32 addr_size_in_bits);
-public:
-        const uint32 size_in_bytes;
-        const uint32 ways;
-        const uint32 line_size;
-        const uint32 addr_size_in_bits;
+    public:
+        LRUModule( std::size_t number_of_sets, std::size_t number_of_ways)
+            : lru_info( number_of_sets, LRUCacheInfo( number_of_ways))
+        { }
+
+        void touch( uint32 num_set, uint32 num_way) { lru_info[ num_set].touch( num_way); }
+        
+        uint32 update( uint32 num_set) { return lru_info[ num_set].update(); }
+
+    private:
+        std::vector<LRUCacheInfo> lru_info;
 };
 
-class CacheTagArray : public CacheTagArrayCheck
+
+
+
+
+
+// Cache tag array module implementation
+class CacheTagArrayCheck : private Log
 {
-    private:
-        /* The set (line) of the cache. */
-        struct CacheTag
-        {
-            Addr line = 0; // data
-            bool is_valid = false; // validity
-        };
+    protected:
+        CacheTagArrayCheck( uint32 size_in_bytes,
+                            uint32 ways,
+                            uint32 size_of_line,
+                            uint32 addr_size_in_bits);
+};
 
-        const uint32 num_sets;
-        const Addr   addr_mask;
 
-        std::vector<std::vector<CacheTag>> array; // array of tags
-        LRUInfo lru; // LRU algorithm module
 
+class CacheTagArray : public CacheTagArrayCheck
+{
     public:
-        CacheTagArray( uint32 size_in_bytes,
+        CacheTagArray( uint32 size_in_bytes, 
                        uint32 ways,
-                       uint32 line_size = 4,
-                       uint32 addr_size_in_bits = 32);
+                       uint32 size_of_line = 4,
+                       uint32 addr_size_in_bits = 32)
+
+            : CacheTagArrayCheck( size_in_bytes, 
+                                  ways,
+                                  size_of_line,
+                                  addr_size_in_bits)
+            , number_of_sets( size_in_bytes / ( ways * size_of_line))
+            , number_of_ways( ways)
```

Comment:
Why were those moved from CacheTagArrayCheck?

Reply:
Why do we need them there? I mean, It's strange to have them in the "Check" class.

Reply:
Denis, that strange kind of discussion is repeating over and over. You are the person who did the changes, so I should be the guy who expects the answers which explain your motivation.

We need it here because we may have it here. "CacheTagArray" has the highest level of abstraction, so if a class was inherited from it we would no more think about correctness of size variables. We may add more methods to the base class which operate with sizes only and then use them for some constant expressions or run-time checks etc. I agree that "Check" may be not the best name, but "CacheTagArraySizeBaseWithBoundaryChecks" has some drawbacks either.

There are _a lot_ of things in C++ programs which may look strange, but they can be even the only one possible solution originated from days and months of the code evolution. If they don't contain a bug, it is better not to touch them.